File: backend/java/src/main/java/ai/snippetquiz/core_service/quiz/domain/events/AIQuestionGeneratedEvent.java
package ai.snippetquiz.core_service.quiz.domain.events;

import ai.snippetquiz.core_service.shared.domain.bus.event.IntegrationEvent;
import ai.snippetquiz.core_service.shared.domain.valueobject.UserId;
import ai.snippetquiz.core_service.shared.domain.Utils;
import lombok.AllArgsConstructor;
<<<<<<< HEAD
import lombok.Builder;
=======
>>>>>>> 363c56c (feat: core service and ai processor event driven)
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

<<<<<<< HEAD
=======
import java.io.Serializable;
>>>>>>> 363c56c (feat: core service and ai processor event driven)
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

@Getter
@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
public class AIQuestionGeneratedEvent extends IntegrationEvent {
    private Integer totalContentEntries;
    private Integer totalContentEntriesSkipped;
    private Integer currentContentEntryIndex;
    private Integer questionsGeneratedSoFar;
    private ContentEntryDto contentEntry;
    private Integer totalChunks;
    private Integer currentChunkIndex;
    private UUID bankId;

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
<<<<<<< HEAD
    @Builder
=======
>>>>>>> 363c56c (feat: core service and ai processor event driven)
    public static class ContentEntryDto {
        private String id;
        private String pageTitle;
        private Integer wordCountAnalyzed;
        private List<QuestionDto> questions;
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    public static class QuestionDto {
        private String question;
        private String type;
        private List<QuestionOptionDto> options;
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    public static class QuestionOptionDto {
        private String optionText;
        private String optionExplanation;
        private Boolean isCorrect;
    }

    public AIQuestionGeneratedEvent(
            UUID aggregateId,
            UserId userId,
            UUID eventId,
            String occurredOn,
            Integer version,
            Integer totalContentEntries,
            Integer totalContentEntriesSkipped,
            Integer currentContentEntryIndex,
            Integer questionsGeneratedSoFar,
            ContentEntryDto contentEntry,
            Integer totalChunks,
            Integer currentChunkIndex,
            UUID bankId) {
        super(aggregateId, userId.getValue(), eventId, occurredOn, version);
        this.totalContentEntries = totalContentEntries;
        this.totalContentEntriesSkipped = totalContentEntriesSkipped;
        this.currentContentEntryIndex = currentContentEntryIndex;
        this.questionsGeneratedSoFar = questionsGeneratedSoFar;
        this.contentEntry = contentEntry;
        this.totalChunks = totalChunks;
        this.currentChunkIndex = currentChunkIndex;
        this.bankId = bankId;
    }

    public static String eventName() {
        return "ai-content-service.questions.generated";
    }

    @Override
    public AIQuestionGeneratedEvent fromPrimitives(
            UUID aggregateId,
            UUID userId,
            HashMap<String, Object> body,
            UUID eventId,
            String occurredOn,
            Integer version) {
        return new AIQuestionGeneratedEvent(
                aggregateId,
                new UserId(userId),
                eventId,
                occurredOn,
                version,
                (Integer) body.get("total_content_entries"),
                (Integer) body.get("total_content_entries_skipped"),
                (Integer) body.get("current_content_entry_index"),
                (Integer) body.get("questions_generated_so_far"),
<<<<<<< HEAD
                Utils.toMap(body.get("content_entry"), ContentEntryDto.class),
=======
                 Utils.toMap(body.get("content_entry"), ContentEntryDto.class),
>>>>>>> 363c56c (feat: core service and ai processor event driven)
                (Integer) body.get("total_chunks"),
                (Integer) body.get("current_chunk_index"),
                UUID.fromString((String) body.get("bank_id")));
    }
}

File: backend/nestjs/apps/ai-processor/src/quiz/events/ai-question-generated.event.ts
import { DomainEvent, DomainEventAttributes } from '../../../../commons/event-bus/domain-event';

export interface QuestionOptionDto {
    optionText: string;
    optionExplanation: string;
    isCorrect: boolean;
}

export interface QuestionDto {
    question: string;
    type: string;
    options: QuestionOptionDto[];
}

export interface ContentEntryDto {
    id: string;
    pageTitle: string | null;
    wordCountAnalyzed: number | null;
    questions: QuestionDto[];
}

export class AIQuestionGeneratedEvent extends DomainEvent {
    static readonly EVENT_NAME = 'ai-processor.questions.generated';

    readonly totalContentEntries: number;
    readonly totalContentEntriesSkipped: number;
    readonly currentContentEntryIndex: number;
    readonly questionsGeneratedSoFar: number;
    readonly contentEntry: ContentEntryDto;
    readonly totalChunks: number;
    readonly currentChunkIndex: number;
    readonly bankId: string;

    constructor(
        aggregateId: string,
        userId: string,
        totalContentEntries: number,
        totalContentEntriesSkipped: number,
        currentContentEntryIndex: number,
        questionsGeneratedSoFar: number,
        contentEntry: ContentEntryDto,
        totalChunks: number,
        currentChunkIndex: number,
        bankId: string,
        eventId?: string,
        occurredOn?: string,
    ) {
        super(
            AIQuestionGeneratedEvent.EVENT_NAME,
            aggregateId,
            userId,
            eventId,
            occurredOn,
        );
        this.totalContentEntries = totalContentEntries;
        this.totalContentEntriesSkipped = totalContentEntriesSkipped;
        this.currentContentEntryIndex = currentContentEntryIndex;
        this.questionsGeneratedSoFar = questionsGeneratedSoFar;
        this.contentEntry = contentEntry;
        this.totalChunks = totalChunks;
        this.currentChunkIndex = currentChunkIndex;
        this.bankId = bankId;
    }

    toPrimitives(): DomainEventAttributes {
        return {
            total_content_entries: this.totalContentEntries,
            total_content_entries_skipped: this.totalContentEntriesSkipped,
            current_content_entry_index: this.currentContentEntryIndex,
            questions_generated_so_far: this.questionsGeneratedSoFar,
            content_entry: this.contentEntry,
            total_chunks: this.totalChunks,
            current_chunk_index: this.currentChunkIndex,
            bank_id: this.bankId,
            aggregate_id: this.aggregateId,
            user_id: this.userId,
        };
    }

    static fromPrimitives(
        body: DomainEventAttributes,
        eventId: string,
        occurredOn: string,
    ): AIQuestionGeneratedEvent {
        return new AIQuestionGeneratedEvent(
            body.aggregate_id,
            body.user_id,
            body.total_content_entries,
            body.total_content_entries_skipped,
            body.current_content_entry_index,
            body.questions_generated_so_far,
            body.content_entry as ContentEntryDto,
            body.total_chunks,
            body.current_chunk_index,
            body.bank_id,
            eventId,
            occurredOn,
        );
    }
}
